### PR TITLE
Gandlf MLCube documentation

### LINKED ISSUE #399: documentation of Building an mlcube from GaNDLF

### PR DESCRIPTION
**Pull request by hasan7n**
This PR creates a hands-on tutorial for creating a GaNDLF MLCube ready to be used for MedPerf. Closes #399 

Preview: https://hasan7n.github.io/medperf/mlcubes/gandlf_mlcube/

### CODE REVIEW

**Comment by hasan7n** [comment]
TODO: 
1. reorganize the MLCubes section in the documentation (create a folder "from_scratch"??)
2. refer readers in the model owner demo to this
3. check what should be moved to gandlf repo?

**Review by sarthakpati — APPROVED** [review]
LGTM

**Review by aristizabal95 — CHANGES_REQUESTED** [review]
I'm getting some errors when trying to train the GaNDLF model:
1. It seems the `config_classification.yaml` file has a syntax error. Line 108 is missing a colon
Currently:
```yaml
'noise': { # for details, see https://torchio.readthedocs.io/transforms/augmentation.html?highlight=randomblur#torchio.transforms.RandomNoise
      'mean': 0 # default mean
      'std': [0, 1] # default std-dev range
    },
```
Expected:
```yaml
'noise': { # for details, see https://torchio.readthedocs.io/transforms/augmentation.html?highlight=randomblur#torchio.transforms.RandomNoise
      'mean': 0, # default mean
      'std': [0, 1] # default std-dev range
    },
```
After fixing that I'm still getting the following error:
```
(venv_gandlf) 14:18:26:mlcommons/medperf-2/medperf_tutorial % gandlf_run \
  -c ./config_classification.yaml \
  -i ./inputs.csv \
  -m ./trained_model_output \
  -t True \
  -d cpu

The ``converters`` are currently experimental. It may not support operations including (but not limited to) Functions in ``torch.nn.functional`` that involved data dimension
WARNING: Initializing 'norm_type' as 'batch'
WARNING: 'batch_norm' is no longer supported, please use 'norm_type' in 'model' instead
WARNING: Initializing 'verbose' as False
WARNING: Initializing 'medcam_enabled' as False
WARNING: Initializing 'save_training' as False
WARNING: Initializing 'pin_memory_dataloader' as False
WARNING: Initializing 'enable_padding' as False
WARNING: Initializing 'scaling_factor' as 1
WARNING: Initializing 'clip_grad' as None
WARNING: Initializing 'track_memory_usage' as False
WARNING: Initializing 'print_rgb_label_warning' as True
WARNING: Initializing 'grid_aggregator_overlap' as crop
WARNING: Initializing 'determinism' as False
WARNING: Initializing 'clip_mode' as None
WARNING: Setting default step_size to: 0.0002
Traceback (most recent call last):
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/bin/gandlf_run", line 129, in <module>
    main_run(
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/GANDLF/cli/main_run.py", line 109, in main_run
    TrainingManager(
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/GANDLF/training_manager.py", line 73, in TrainingManager
    dataframe[dataframe.columns[parameters["headers"]["subjectIDHeader"]]]
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/pandas/core/frame.py", line 2912, in __getitem__
    indexer = self.loc._get_listlike_indexer(key, axis=1, raise_missing=True)[1]
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/pandas/core/indexing.py", line 1254, in _get_listlike_indexer
    self._validate_read_indexer(keyarr, indexer, axis, raise_missing=raise_missing)
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/pandas/core/indexing.py", line 1298, in _validate_read_indexer
    raise KeyError(f"None of [{key}] are in the [{axis_name}]")
KeyError: "None of [Index([('001', '2d_rad_segmentation/001/image.png', '2d_rad_segmentation/001/mask.png')], dtype='object')] are in the [columns]"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/bin/gandlf_run", line 140, in <module>
    sys.exit("Error in GANDLF run:", e)
TypeError: exit expected at most 1 argument, got 2
```
I believe I'm using the latest version of GaNDLF, but let me know if these were the issues you encountered when not.

It still seems that artifacts were created from the previous command, so I decided to keep going. It appears that when running `gandlf_deploy`, MLCube is trying to get the image from the docker hub and failing. Did you encounter this? This is the traceback:
```
(venv_gandlf) 15:02:55:mlcommons/medperf-2/medperf_tutorial % gandlf_deploy \             
  -c ./config_classification.yaml \
  -m ./trained_model_output \
  --target docker \
  --mlcube-root ./my_gandlf_mlcube \
  -o ./built_gandlf_mlcube

The ``converters`` are currently experimental. It may not support operations including (but not limited to) Functions in ``torch.nn.functional`` that involved data dimension
Connected to the docker service.
Running MLCube configuration with the following command:
mlcube_docker configure --platform=docker  -Prunner.build_strategy=always --mlcube=/Users/alejandroaristizabal/Documents/mlcommons/medperf-2/medperf_tutorial/my_gandlf_mlcube -Prunner.build_context=/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages
If this is your first GaNDLF deployment, this may take longer than usual while image layers are built.
2023-04-25 15:03:18 FACTORED-NWRQCDVJ44 mlcube.config[43710] WARNING Default runner config contains parameters that are not present in the effective runner config (params=['--network', '--security-opt', '--gpus', '--memory', '--cpuset-cpus']). This probably means that a new version of a runner was installed that introduced new parameters.
Error response from daemon: pull access denied for johnsmith/gandlf_model, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
2023-04-25 15:03:19 FACTORED-NWRQCDVJ44 mlcube.shell[43710] ERROR Command='docker pull johnsmith/gandlf_model:0.0.1' status=256 exit_status=exited exit_code=1 on_error=raise
2023-04-25 15:03:19 FACTORED-NWRQCDVJ44 mlcube.__main__[43710] ERROR Failed to configure MLCube with error code 1.
Traceback (most recent call last):
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/mlcube_docker/docker_run.py", line 162, in configure
    Shell.run([docker, "pull", image])
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/mlcube/shell.py", line 95, in run
    raise ExecutionError(
mlcube.errors.ExecutionError: Failed to execute shell command.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/mlcube/__main__.py", line 216, in configure
    runner.configure()
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/mlcube_docker/docker_run.py", line 170, in configure
    raise ExecutionError.mlcube_configure_error(
mlcube.errors.ExecutionError: DockerRun runner failed to configure MLCube. Error occurred while pulling docker image (docker=docker, image=johnsmith/gandlf_model:0.0.1).
ERROR:
        message: DockerRun runner failed to configure MLCube.
        description: Error occurred while pulling docker image (docker=docker, image=johnsmith/gandlf_model:0.0.1).
        context: {'status': 'exited', 'code': 1, 'cmd': 'docker pull johnsmith/gandlf_model:0.0.1'}
Traceback (most recent call last):
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/bin/gandlf_deploy", line 83, in <module>
    result = run_deployment(
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/GANDLF/cli/deploy.py", line 51, in run_deployment
    result = deploy_docker_mlcube(
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/GANDLF/cli/deploy.py", line 170, in deploy_docker_mlcube
    assert (
AssertionError: mlcube_docker configuration failed. Check output for more details.
```

The instructions look great! I think we should look into being able to fill in the cookiecutter variables through inline parameters. My lazy self went ahead and used the defaults, which then had me confused for a few minutes into why the other command was not able to find the mlcube in question haha!

**Comment by hasan7n** [comment]
@aristizabal95 @sarthakpati It must be a misleading instruction or an error in the setup. I ran it successfully but before writing.

**Comment by hasan7n** [comment]
@aristizabal95 I am curious: what `pandas` version do you have when run this?

**Comment by aristizabal95** [comment]
I'll go through the whole tutorial another time to see if I messed up some configuration. Something else I forgot to mention was that running the command for installing GaNDLF dependencies for CPU fails with the following message:
```
pip install torch==1.13.1+cpu torchvision==0.14.1+cpu torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cpu
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cpu
ERROR: Could not find a version that satisfies the requirement torch==1.13.1+cpu (from versions: 1.4.0, 1.5.0, 1.5.1, 1.6.0, 1.7.0, 1.7.1, 1.8.0, 1.8.1, 1.9.0, 1.9.1, 1.10.0, 1.10.1, 1.10.2, 1.11.0, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 2.0.0)
ERROR: No matching distribution found for torch==1.13.1+cpu
```
My intuition was to remove the `+cpu` from torch and it executed correctly. Could this cause issues?

**Comment by aristizabal95** [comment]
@hasan7n Here's my pip version: 
```
pip 23.0.1 from /Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/pip (python 3.8)
```
Also, the installed version of pandas is:
```
pandas==1.1.5
```
@sarthakpati I'm trying to reinstall GaNDLF but I'm getting this message:
```
% pip install gandlf
ERROR: Could not find a version that satisfies the requirement gandlf (from versions: 0.0.1, 0.0.16.dev0, 0.0.16.dev20230321, 0.0.16.dev20230322, 0.0.16.dev20230323, 0.0.16.dev20230324, 0.0.16.dev20230325, 0.0.16.dev20230326, 0.0.16.dev20230327, 0.0.16.dev20230328, 0.0.16.dev20230329, 0.0.16.dev20230330, 0.0.16.dev20230331, 0.0.16.dev20230401, 0.0.16.dev20230402, 0.0.16.dev20230403, 0.0.16.dev20230404, 0.0.16.dev20230405, 0.0.16.dev20230406, 0.0.16.dev20230407, 0.0.16.dev20230408, 0.0.16.dev20230409, 0.0.16.dev20230410, 0.0.16.dev20230411, 0.0.16.dev20230412, 0.0.16.dev20230413, 0.0.16.dev20230414, 0.0.16.dev20230415, 0.0.16.dev20230416, 0.0.16.dev20230417, 0.0.16.dev20230418, 0.0.16.dev20230419, 0.0.16.dev20230420, 0.0.16.dev20230421, 0.0.16.dev20230422, 0.0.16.dev20230423, 0.0.16.dev20230424, 0.0.16.dev20230425, 0.0.16.dev20230426)
ERROR: No matching distribution found for gandlf
```
I got that too last time, and what I did was install the nightly version with `pip install --pre gandlf`. Could that be the cause of having the outdated version of `config_classification.yaml`?

**Comment by aristizabal95** [comment]
I suggest modifying the **Before we start** script so that it includes the step to `cd` to the respective directory:
 ```
 sh tutorials_scripts/setup_GaNDLF_mlcube_tutorial.sh
cd medperf_tutorial
 ```
 While also keep reminding the user where we expect the other commands to happen. This way even someone skimming through the tutorial will be able to run the code.

**Comment by aristizabal95** [comment]
I'm getting an error when trying to use `gandlf_deploy`
```
% gandlf_deploy \
  -c ./config_classification.yaml \
  -m ./trained_model_output \
  --target docker \
  --mlcube-root ./my_gandlf_mlcube \
  -o ./built_gandlf_mlcube

The ``converters`` are currently experimental. It may not support operations including (but not limited to) Functions in ``torch.nn.functional`` that involved data dimension
Connected to the docker service.
Running MLCube configuration with the following command:
mlcube_docker configure --platform=docker  -Prunner.build_strategy=always --mlcube=/Users/alejandroaristizabal/Documents/mlcommons/medperf-2/medperf_tutorial/my_gandlf_mlcube -Prunner.build_context=/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages
If this is your first GaNDLF deployment, this may take longer than usual while image layers are built.
Error response from daemon: pull access denied for johnsmith/gandlf_model, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
2023-05-03 10:50:45 FACTORED-NWRQCDVJ44 mlcube.shell[65722] ERROR Command='docker pull johnsmith/gandlf_model:0.0.1' status=256 exit_status=exited exit_code=1 on_error=raise
2023-05-03 10:50:45 FACTORED-NWRQCDVJ44 mlcube.__main__[65722] ERROR Failed to configure MLCube with error code 1.
Traceback (most recent call last):
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/mlcube_docker/docker_run.py", line 162, in configure
    Shell.run([docker, "pull", image])
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/mlcube/shell.py", line 95, in run
    raise ExecutionError(
mlcube.errors.ExecutionError: Failed to execute shell command.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/mlcube/__main__.py", line 216, in configure
    runner.configure()
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/mlcube_docker/docker_run.py", line 170, in configure
    raise ExecutionError.mlcube_configure_error(
mlcube.errors.ExecutionError: DockerRun runner failed to configure MLCube. Error occurred while pulling docker image (docker=docker, image=johnsmith/gandlf_model:0.0.1).
ERROR:
        message: DockerRun runner failed to configure MLCube.
        description: Error occurred while pulling docker image (docker=docker, image=johnsmith/gandlf_model:0.0.1).
        context: {'status': 'exited', 'code': 1, 'cmd': 'docker pull johnsmith/gandlf_model:0.0.1'}
Traceback (most recent call last):
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/bin/gandlf_deploy", line 83, in <module>
    result = run_deployment(
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/GANDLF/cli/deploy.py", line 51, in run_deployment
    result = deploy_docker_mlcube(
  File "/Users/alejandroaristizabal/opt/anaconda3/envs/venv_gandlf/lib/python3.8/site-packages/GANDLF/cli/deploy.py", line 176, in deploy_docker_mlcube
    assert (
AssertionError: mlcube_docker configuration failed. Check output for more details.
```

MLCube is trying to pull the image from docker hub, which it can since the image doesn't exist, and the user is not the one I'm logged with (`johnsmith`) Is there a way to make the `gandlf_deploy` fallback to building the image if it fails when trying to retrieve it?

File: docs/mlcubes/gandlf_mlcube.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+We assume that you already have [medperf installed](../installation.md) and [GaNDLF installed](https://mlcommons.github.io/GaNDLF/setup/).
+
+## Before We Start
+
+#### Download the Necessary files
+
+We provide a script that downloads necessary files so that you follow the tutorial smoothly. Run the following: (make sure you are in MedPerf's root folder)
+
+```bash
+sh tutorials_scripts/setup_GaNDLF_mlcube_tutorial.sh
```
Comment (aristizabal95) [inline docs/mlcubes/gandlf_mlcube.md:20]: I suggest adding `cd medperf_tutorial` after this

**Review by aristizabal95 — CHANGES_REQUESTED** [review]
I gave a small suggestion and got some errors when trying to run the tutorial. Give them a look and let me know what you think!

**Comment by hasan7n** [comment]
I see. Looking [here](https://pytorch.org/get-started/previous-versions/) shows that installing pytorch on MAC is different. We can open an issue on gandlf repo for this if are officially supporting MAC.

However I am wondering: why should it fail in the linux docker image?

**Pull request merged**